src/main/java/org/folio/services/financedata/FinanceDataService.java

```diff
@@ -1,39 +1,206 @@
+    log.debug("Trying to update finance data collection with size: {}", financeDataCollection.getTotalRecords());
+    if (CollectionUtils.isEmpty(financeDataCollection.getFyFinanceData())) {
+      log.info("Finance data collection is empty, nothing to update");
+      return succeededFuture();
+    }
+
+    validateFinanceDataCollection(financeDataCollection, getFiscalYearId(financeDataCollection));
+
+    return processAllocationTransaction(financeDataCollection, requestContext)
+      .compose(v -> updateFinanceData(financeDataCollection, requestContext))
+      .onSuccess(asyncResult -> processLogs(financeDataCollection, requestContext, COMPLETED))
+      .onFailure(asyncResult -> processLogs(financeDataCollection, requestContext, ERROR));
+  }
+
+  private void validateFinanceDataCollection(FyFinanceDataCollection financeDataCollection, String fiscalYearId) {
+    for (int i = 0; i < financeDataCollection.getFyFinanceData().size(); i++) {
+      var financeData = financeDataCollection.getFyFinanceData().get(i);
+      validateFinanceDataFields(financeData, i, fiscalYearId);
+
+      var allocationChange = BigDecimal.valueOf(financeData.getBudgetAllocationChange());
+      var initialAllocation = BigDecimal.valueOf(financeData.getBudgetInitialAllocation());
+
+      if (allocationChange.doubleValue() < 0 && allocationChange.abs().doubleValue() > initialAllocation.doubleValue()) {
+        var error = createError("Allocation change cannot be greater than initial allocation",
+          String.format("financeData[%s].budgetAllocationChange", i), String.valueOf(financeData.getBudgetAllocationChange()));
+        throw new HttpException(422, new Errors().withErrors(List.of(error)));
+      }
+    }
+  }
+
+  private Future<Void> processAllocationTransaction(FyFinanceDataCollection fyFinanceDataCollection,
+                                                    RequestContext requestContext) {
+    return fiscalYearService.getFiscalYearById(getFiscalYearId(fyFinanceDataCollection), requestContext)
+      .map(fiscalYear -> createAllocationTransactions(fyFinanceDataCollection, fiscalYear.getCurrency()))
+      .compose(transactions -> createBatchTransaction(transactions, requestContext));
+  }
+
+  private String getFiscalYearId(FyFinanceDataCollection fyFinanceDataCollection) {
+    return fyFinanceDataCollection.getFyFinanceData().get(0).getFiscalYearId();
+  }
+
+  private List<Transaction> createAllocationTransactions(FyFinanceDataCollection financeDataCollection, String currency) {
+    return financeDataCollection.getFyFinanceData().stream()
+      .map(financeData -> createAllocationTransaction(financeData, currency))
+      .toList();
+  }
+
+  private Transaction createAllocationTransaction(FyFinanceData financeData, String currency) {
+    var allocation = calculateAllocation(financeData);
+    log.info("Creating allocation transaction for fund '{}' and budget '{}' with allocation '{}'",
```

Comment:
Here can be included too: createAllocationTransaction::

Reply:
thanks updated